### PR TITLE
Create Azure Bicep version.

### PR DESCRIPTION
**Pull request by mcollier**
# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

- Adds a **main.bicep** file to implement Azure Functions with private endpoints for the backend Azure Storage account.
- Sets **azuredeploy.json** to be the output from the Azure Bicep compile operation.
- Updates the architecture diagram.
- Fixes a minor typo in the **metadata.json** file, and updates the date.

### CODE REVIEW

**Comment by mcollier** [comment]
Would you please give this a review, @gabesmsft?

File: quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+  sku: {
+    name: 'Standard_LRS'
+  }
+  properties: {
+    networkAcls: {
+      bypass: 'None'
+      defaultAction: 'Deny'
+    }
+  }
+  dependsOn: [
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep:337]: why is this needed?

File: quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+    virtualNetwork
+  ]
+}
+
+resource functionContentShare 'Microsoft.Storage/storageAccounts/fileServices/shares@2021-04-01' = {
+  name: '${storageAccount.name}/default/${functionContentShareName}'
+  dependsOn: [
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep:344]: use the parent property and remove the dependsOn

File: quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+    family: 'EP'
+  }
+  kind: 'elastic'
+  properties: {
+    maximumElasticWorkerCount: 20
+    reserved: isReserved
+  }
+}
+
+resource function 'Microsoft.Web/sites@2021-01-01' = {
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep:374]: function is pretty generic - let's use "functionApp" or something more descriptive

File: quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+          value: '1'
+        }
+        {
+          name: 'WEBSITE_CONTENTSHARE'
+          value: functionContentShareName
+        }
+      ]
+    }
+  }
+  dependsOn: [
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep:420]: move dependsOn above properties

**Review by bmoore-msft — CHANGES_REQUESTED** [review]
also check the pipeline for the BP changes needed

File: quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+param location string = resourceGroup().location
+
+@description('The name of the Azure Function app.')
+param functionAppName string = 'func-${uniqueString(resourceGroup().id)}'
+
+@description('The name of the Azure Function hosting plan.')
+param functionAppPlanName string = 'plan-${uniqueString(resourceGroup().id)}'
+
+@description('Determines if a Windows or Linux-based hosting plan is created.  Set to true for Linux, and false for Windows.')
+param isReserved bool = false
```
[the remark this answers is not in the document]
Comment (mcollier) [reply quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep:11]: The "reserved" term comes from the Microsoft.Web/serverFarm resource - https://docs.microsoft.com/en-us/azure/templates/microsoft.web/serverfarms?tabs=json#appserviceplanproperties-object. I'll change the parameter description to more closely match what's on the documentation for the "reserved" property.

[the remark this answers is not in the document]
Comment (bmoore-msft) [reply quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep:11]: Got it - for the user, is the user expected to pick Linux vs. Windows?  Or reserved vs. not?  My guess it user doesn't care about the implementation of the web app, just whether they want linux v. windows.  And if that's accurate, then we should probably rename the parameter accordingly...

File: quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+      }
+    ]
+  }
+}
+
+// -- Private DNS Zones --
+resource storageFileDnsZone 'Microsoft.Network/privateDnsZones@2020-06-01' = {
+  name: privateStorageFileDnsZoneName
+  location: 'global'
+  dependsOn: [
```
[the remark this answers is not in the document]
Comment (mcollier) [reply quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep]: Removed.

File: quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep (hunk 1 above)
Comment (mcollier) [reply quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep]: Removed.

File: quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep (hunk 3 above)
Comment (mcollier) [reply quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep]: Done.

File: quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep (hunk 4 above)
Comment (mcollier) [reply quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep]: Done.

File: quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep (hunk 5 above)
[the remark this answers is not in the document]
Comment (mcollier) [reply quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep:11]: The term "reserved" is a poor name to indicate if the function app is based on Windows or Linux.  If reserved, that means use Linux, otherwise use Windows.

Do you think reworking to something like "functionPlanOS" with allowed values of "Windows" and "Linux" would be easier?  If so, I could probably rework some of the conditional logic in the .bicep file to change the necessary values if the value is Windows or Linux.

File: quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep (hunk 2 above)
Comment (mcollier) [reply quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep]: I removed 'dependsOn'.  I don't think `parent` is needed either.  The  **azuredeploy.json** file, resulting from `az bicep build -f` shows the dependency gets added to the ARM.

image

File: quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep (hunk 5 above)
[the remark this answers is not in the document]
Comment (bmoore-msft) [reply quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep:11]: yeah I think your logic is more or less the same regardless of the name... but having user choose Windows/Linux seems to be the right pivot.  functionPlanOS seems fine

File: quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep (hunk 2 above)
Comment (bmoore-msft) [reply quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep]: LGTM

**Review by bmoore-msft — CHANGES_REQUESTED** [review]
I think the only think left is the param name for the OS type - rest of the changes look good.

File: quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep [hunk 7]
```diff
@@ -8,7 +8,11 @@ param functionAppName string = 'func-${uniqueString(resourceGroup().id)}'
 param functionAppPlanName string = 'plan-${uniqueString(resourceGroup().id)}'
 
 @description('If Linux app service plan, set to true, false otherwise (for Windows).')
-param isReserved bool = false
+@allowed([
+  'Windows'
+  'Linux'
+])
+param functionPlanOS string = 'Windows'
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/function-app-storage-private-endpoints/main.bicep:15]: description needs to be updated ;)

**Review by bmoore-msft — CHANGES_REQUESTED** [review]
